File: bot-api/src/main/java/gg/mineral/bot/api/world/ClientWorld.java
package gg.mineral.bot.api.world;

import gg.mineral.bot.api.entity.ClientEntity;
import gg.mineral.bot.api.world.block.Block;

import java.util.Collection;

public interface ClientWorld {

    /**
     * Gets all entities in the world.
     *
     * @return all entities in the world
     */
    Collection<ClientEntity> getEntities();

    /**
     * Gets the entity with the specified ID.
     *
     * @param entityId the entity ID
     * @return the entity with the specified ID
     */
    ClientEntity getEntityByID(int entityId);

    /**
     * Gets the block at the specified coordinates.
     *
     * @param x the x-coordinate
     * @param y the y-coordinate
     * @param z the z-coordinate
     * @return the block at the specified coordinates
     */
    Block getBlockAt(int x, int y, int z);

    /**
     * Gets the block at the specified coordinates.
     *
     * @param x the x-coordinate
     * @param y the y-coordinate
     * @param z the z-coordinate
     * @return the block at the specified coordinates
     */
    Block getBlockAt(double x, double y, double z);
}

File: bot-api/src/main/kotlin/gg/mineral/bot/api/world/ClientWorld.kt
package gg.mineral.bot.api.world

import gg.mineral.bot.api.entity.ClientEntity
import gg.mineral.bot.api.world.block.Block

interface ClientWorld {
    /**
     * Gets all entities in the world.
     *
     * @return all entities in the world
     */
    val entities: Collection<ClientEntity>

    /**
     * Gets the entity with the specified ID.
     *
     * @param entityId the entity ID
     * @return the entity with the specified ID
     */
    fun getEntityByID(entityId: Int): ClientEntity?

    /**
     * Gets the block at the specified coordinates.
     *
     * @param x the x-coordinate
     * @param y the y-coordinate
     * @param z the z-coordinate
     * @return the block at the specified coordinates
     */
    fun getBlockAt(x: Int, y: Int, z: Int): Block

    /**
     * Gets the block at the specified coordinates.
     *
     * @param x the x-coordinate
     * @param y the y-coordinate
     * @param z the z-coordinate
     * @return the block at the specified coordinates
     */
    fun getBlockAt(x: Double, y: Double, z: Double): Block

    /**
     * Creates a deep copy of this world.
     *
     * @return a deep copy of this world
     */
    fun deepCopy(): ClientWorld
}
